adaptive_memory_v4.0.py

```
@@ -0,0 +1,2095 @@
+                # Use text-based similarity - no embedding to cache
+                is_dup = await self._check_text_similarity(text, all_memories, exclude_id=exclude_id)
+                return is_dup, None
+                        
+            return False, new_embedding if self.valves.use_embeddings_for_deduplication else None  # Not duplicate, return embedding for caching
+            
+        except Exception as e:
+            logger.error(f"Error during duplicate check: {e}")
+            logger.error(f"Traceback: {traceback.format_exc()}")
+            # If deduplication fails, err on the side of caution and keep the memory
+            return False, None
+    async def _check_text_similarity(self, text: str, all_memories: List[Any], exclude_id: str = None) -> bool:
```

Ruff (0.14.10)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)